hazelcast/src/main/java/com/hazelcast/map/impl/event/MapEventPublisherImpl.java

```diff
@@ -140,75 +158,106 @@ public void publishEvent(Address caller, String mapName, EntryEventType eventTyp
             return;
         }
 
-        List<EventRegistration> includeValueRegistrations = null;
-        List<EventRegistration> nullValueRegistrations = null;
+        publishEvent(registrations, caller, mapName, eventType, dataKey, oldValue, value, mergingValue);
+    }
+
+    private void publishEvent(Collection<EventRegistration> registrations, Address caller, String mapName,
+                                     EntryEventType eventType, Data dataKey, Object oldValue, Object value,
+                                     Object mergingValue) {
+
+        Map<Integer, EntryEventData> eventDataIncludingValues = new Int2ObjectHashMap<EntryEventData>(EVENT_DATA_MAP_CAPACITY);
+        Map<Integer, EntryEventData> eventDataExcludingValues = new Int2ObjectHashMap<EntryEventData>(EVENT_DATA_MAP_CAPACITY);
```

Comment:
We can defer `Int2ObjectHashMap` creation at a later point, and can do it when needed, now for ex., if i have `value-included` event registrations and don't have any `value-excluded` ones, there will be always two `Int2ObjectHashMap` instance creation although one is sufficient. This event publishing part is on hot-path and it is better to create possible minimum number of objects.